### PR TITLE
Kills /obj/item/melee/transforming, replaces it with a transforming weapon component

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

This PR kills off the `transforming` subtype of `/obj/item/melee` and replaces it with a component to handle the transforming behavior, `/datum/component/transforming`.

The transforming component handles updating the variables of an item when it's transformed. Things like force, sharpness, whetstone force bonus, and attack verbs. Similar to the two-handed component, but instead of transforming into a two-hander it remains a one handed weapon.

The "nemesis" behavior (dealing addition damage to certain factions) of the transforming subtype was moved to the cleaving saw only, since it was the only transforming item that used it. In the future, this can be made into a bespoke element/component as well.

The following weapons and items have been updated to use this component:
- Energy Swords / Sabers / Bananium Energy Sword
- Energy Circular Saw
- Energy Dagger
- Energy Axe
- Toy Energy Sword
- Holographic Energy Sword
- Switchblade
- Advanced Medical Tools (Laser scalpel, Mechanical Pinches, Searing Tool)
- Advanced Engineering Tools (Hand Drill, Jaws of Life / Syndicate Jaws of Life)
- Combat Wrench
- Cleaving Saw
- Telescopic Batons / Contractor Batons
- Roasting Stick
- Telescopic Riot Shield
- Energy Shield / Bananium Energy Shield

This PR also touches up the code around the various above items.

## Why It's Good For The Game

Someone asked me how they'd make their nullrod reskin transform like an energy weapon, and I realized there was no good way besides copy + pasting a ton of code.

Instead of copy + pasting the same `attack_self` sequence of transforming across multiple items, this allows for them all to behave semi-consistently. Which is neat.

## Changelog
:cl: Melbert
refactor: Various toggle/transforming items (energy swords / shields, the cleaving saw, switchblades - to name a few) are now handled via component (the transforming component).
code: Slightly cleaned up the code surrounding various transforming items. Comments, better vars, etc.
fix: The Syndicate Jaws of Life no longer become normal Jaws of Life on use
/:cl:

### CODE REVIEW

File: code/datums/components/transforming.dm [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+	/// List of continuous attack verbs used when the weapon is enabled
+	var/list/attack_verb_continuous_on = list("attacks", "slashes", "stabs", "slices", "tears", "lacerates", "rips", "dices", "cuts")
+	/// List of simple attack verbs used when the weapon is enabled
+	var/list/attack_verb_simple_on = list("attack", "slash", "stab", "slice", "tear", "lacerate", "rip", "dice", "cut")
+	/// Whether clumsy people need to succeed an RNG check to turn it on without hurting themselves
+	var/clumsy_check = TRUE
+	/// If we get sharpened with a whetstone, save the bonus here for later use if we un/redeploy
+	var/sharpened_bonus
+	/// Callback to be invoked whenever the weapon is transformed.
+	var/datum/callback/on_transform_callback
```
Comment (tralezab) [inline code/datums/components/transforming.dm:34]: move all the default values above here...

File: code/datums/components/transforming.dm [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+		transform_cooldown_time,
+		force_on,
+		throwforce_on,
+		sharpness_on,
+		hitsound_on,
+		w_class_on,
+		clumsy_check = TRUE,
+		list/attack_verb_continuous_on,
+		list/attack_verb_simple_on,
+		datum/callback/on_transform_callback,
```
Comment (tralezab) [inline code/datums/components/transforming.dm:49]: onto the initial arguments! makes it easier for admins to add "default" transforming items, and even if broken, you don't have to check if you're replacing stuff. continuing with this thought...

File: code/datums/components/transforming.dm [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.transform_cooldown_time = transform_cooldown_time
+	src.force_on = force_on
+	src.throwforce_on = throwforce_on
+	src.sharpness_on = sharpness_on
+	src.hitsound_on = hitsound_on
+	src.w_class_on = w_class_on
+	src.clumsy_check = clumsy_check
```
Comment (tralezab) [inline code/datums/components/transforming.dm:61]: you're going to be replacing src values with null without the default values being sent.

File: code/datums/components/transforming.dm [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	if(islist(attack_verb_continuous_on))
+		src.attack_verb_continuous_on = attack_verb_continuous_on
+	if(islist(attack_verb_simple_on))
+		src.attack_verb_simple_on = attack_verb_simple_on
```
Comment (tralezab) [inline code/datums/components/transforming.dm:65]: you're having to check islist without the default values being sent.

**Review by tralezab — COMMENTED** [review]
Here's an example of the definitions given to the component being in the initialize (or in case of element, attach)

https://github.com/tgstation/tgstation/blob/e7f846bce2498ef03810870dd12d32b4b638af82/code/datums/components/faction_granter.dm#L18

**Comment by Fikou** [comment]
holo eswords should use it unless you added them already
also if you update the laser scalpel you should update the rest of the power tools

**Comment by necromanceranne** [comment]
Could this get added to the edagger as well?

**Comment by MrMelbert** [comment]
sure, I by all means have no exhaustive list of things that transform so shout them out as you find them

**Comment by Twaticus** [comment]
icons/obj/tools.dmi - the syndie jaws cutter icon seems to be missing 
image

**Comment by MrMelbert** [comment]
> 
> icons/obj/tools.dmi - the syndie jaws cutter icon seems to be missing

i don't see it, i'll double check but I think it's just a github issue

image

**Comment by Twaticus** [comment]
damn youre right its there. i fell for the icondiffbot ruse

**Review by Twaticus — APPROVED** [review]

File: code/datums/components/transforming.dm [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+/datum/component/transforming/proc/on_attack_self(obj/item/source, mob/user)
+	SIGNAL_HANDLER
+
+	if(!COOLDOWN_FINISHED(src, transform_cooldown))
+		to_chat(user, span_warning("Wait a bit before trying to use [source] again!"))
+		return
+
+	if(pre_transform_callback)
+		if(!pre_transform_callback.Invoke(user))
+			return
```
Comment (LemonInTheDark) [inline code/datums/components/transforming.dm:125]: You could do this with signals, but this is alright too, just a bit more limiting

File: code/datums/components/transforming.dm [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+ * *
+ * source - the item being transformed / parent
+ * user - the mob transforming the item
+ *
+ * returns TRUE.
+ */
+/datum/component/transforming/proc/do_transform(obj/item/source, mob/user)
+	toggle_active(source)
+	if(on_transform_callback)
+		on_transform_callback.Invoke(user, active)
```
Comment (LemonInTheDark) [inline code/datums/components/transforming.dm:145]: kinda ditto with this

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/mining/lavaland/megafauna_loot.dm [hunk 7]
```diff
@@ -795,101 +795,124 @@
...
 		var/static/list/cleaving_saw_cleave_angles = list(0, -45, 45) //so that the animation animates towards the target clicked and not towards a side target
 		for(var/i in cleaving_saw_cleave_angles)
-			var/turf/turf = get_step(user_turf, turn(dir_to_target, i))
-			for(var/mob/living/living_target in turf)
-				if(user.Adjacent(living_target) && living_target.body_position != LYING_DOWN)
-					melee_attack_chain(user, living_target)
+			var/turf/T = get_step(user_turf, turn(dir_to_target, i))
+			for(var/mob/living/L in T)
+				if(user.Adjacent(L) && L.density)
+					melee_attack_chain(user, L)
```
[the remark this answers is not in the document]
Comment (MrMelbert) [reply code/modules/mining/lavaland/megafauna_loot.dm:886]: whoops, accidentally reverted when the files were moved

File: code/datums/components/transforming.dm (hunk 5 above)
Comment (MrMelbert) [reply code/datums/components/transforming.dm]: Sure

**Pull request merged**